### PR TITLE
Support for Huawei SmartAX

### PR DESCRIPTION
**Pull request by BSpendlove**
You know I've been working on this, I've seen someone send a pull request for a telnet version of the OLTs so I thought I'd get in here quick and get this kicked off. I think it should be implemented in a seperate file (not the `huawei.py`) because it is technically a completely different operating system that acts differently to the OS based on the Huawei routers.

Almost every check passes, however I'm having issues with the 2nd SSH that occurs during the `test_netmiko_show.py` because you are only allowed to have a single session via SSH at any given time with the root account. Please see below the test results and let me know if you want me to change anything:

```
test_netmiko_show.py::test_disable_paging PASSED                         [  5%]
test_netmiko_show.py::test_ssh_connect PASSED                            [ 11%]
test_netmiko_show.py::test_ssh_connect_cm ERROR                          [ 16%]
test_netmiko_show.py::test_send_command_timing PASSED                    [ 22%]
test_netmiko_show.py::test_send_command PASSED                           [ 27%]
test_netmiko_show.py::test_cmd_verify_decorator ERROR                    [ 33%]
test_netmiko_show.py::test_send_command_global_cmd_verify ERROR          [ 38%]
test_netmiko_show.py::test_send_command_juniper SKIPPED                  [ 44%]
test_netmiko_show.py::test_send_command_textfsm SKIPPED                  [ 50%]
test_netmiko_show.py::test_send_command_genie SKIPPED                    [ 55%]
test_netmiko_show.py::test_base_prompt PASSED                            [ 61%]
test_netmiko_show.py::test_strip_prompt PASSED                           [ 66%]
test_netmiko_show.py::test_strip_command PASSED                          [ 72%]
test_netmiko_show.py::test_normalize_linefeeds PASSED                    [ 77%]
test_netmiko_show.py::test_clear_buffer PASSED                           [ 83%]
test_netmiko_show.py::test_enable_mode PASSED                            [ 88%]
test_netmiko_show.py::test_disconnect PASSED                             [ 94%]
test_netmiko_show.py::test_disconnect_no_enable PASSED                   [100%]
```
```
===================================================================================================== short test summary info =====================================================================================================
SKIPPED [1] /home/bspendlove/dev/git/netmiko/tests/test_netmiko_show.py:123: <Skipped instance>
SKIPPED [1] /home/bspendlove/dev/git/netmiko/tests/test_netmiko_show.py:144: TextFSM/ntc-templates not supported on this platform
SKIPPED [1] /home/bspendlove/dev/git/netmiko/tests/test_netmiko_show.py:169: Genie not supported on this platform
============================================================================================ 12 passed, 3 skipped, 3 errors in 42.06s =============================================================================================
```

```
$ py.test -v test_netmiko_config.py --test_device huawei_smartax
======================================================================================================= test session starts =======================================================================================================
platform linux -- Python 3.6.8, pytest-5.4.1, py-1.8.1, pluggy-0.13.1 -- /home/bspendlove/dev/git/venv_netmiko/bin/python3
cachedir: .pytest_cache
rootdir: /home/bspendlove/dev/git/netmiko, inifile: setup.cfg
collected 8 items                                                                                                                                                                                                                 

test_netmiko_config.py::test_ssh_connect PASSED                                                                                                                                                                             [ 12%]
test_netmiko_config.py::test_enable_mode PASSED                                                                                                                                                                             [ 25%]
test_netmiko_config.py::test_config_mode PASSED                                                                                                                                                                             [ 37%]
test_netmiko_config.py::test_exit_config_mode PASSED                                                                                                                                                                        [ 50%]
test_netmiko_config.py::test_config_set PASSED                                                                                                                                                                              [ 62%]
test_netmiko_config.py::test_config_set_longcommand PASSED                                                                                                                                                                  [ 75%]
test_netmiko_config.py::test_config_from_file PASSED                                                                                                                                                                        [ 87%]
test_netmiko_config.py::test_disconnect PASSED                                                                                                                                                                              [100%]

======================================================================================================= 8 passed in 45.37s ========================================================================================================
```

### CODE REVIEW

**Comment by BSpendlove** [comment]
I just went ahead and created a new user account to test and increased the number of logins at one given time which has fixed the 3 errors in my previous `test_netmiko_show.py` py.test, please see below:

```
$ py.test -v test_netmiko_show.py --test_device huawei_smartax
======================================================================================================= test session starts =======================================================================================================
platform linux -- Python 3.6.8, pytest-5.4.1, py-1.8.1, pluggy-0.13.1 -- /home/bspendlove/dev/git/venv_netmiko/bin/python3
cachedir: .pytest_cache
rootdir: /home/bspendlove/dev/git/netmiko, inifile: setup.cfg
collected 18 items                                                                                                                                                                                                                

test_netmiko_show.py::test_disable_paging PASSED                                                                                                                                                                            [  5%]
test_netmiko_show.py::test_ssh_connect PASSED                                                                                                                                                                               [ 11%]
test_netmiko_show.py::test_ssh_connect_cm PASSED                                                                                                                                                                            [ 16%]
test_netmiko_show.py::test_send_command_timing PASSED                                                                                                                                                                       [ 22%]
test_netmiko_show.py::test_send_command PASSED                                                                                                                                                                              [ 27%]
test_netmiko_show.py::test_cmd_verify_decorator PASSED                                                                                                                                                                      [ 33%]
test_netmiko_show.py::test_send_command_global_cmd_verify PASSED                                                                                                                                                            [ 38%]
test_netmiko_show.py::test_send_command_juniper SKIPPED                                                                                                                                                                     [ 44%]
test_netmiko_show.py::test_send_command_textfsm SKIPPED                                                                                                                                                                     [ 50%]
test_netmiko_show.py::test_send_command_genie SKIPPED                                                                                                                                                                       [ 55%]
test_netmiko_show.py::test_base_prompt PASSED                                                                                                                                                                               [ 61%]
test_netmiko_show.py::test_strip_prompt PASSED                                                                                                                                                                              [ 66%]
test_netmiko_show.py::test_strip_command PASSED                                                                                                                                                                             [ 72%]
test_netmiko_show.py::test_normalize_linefeeds PASSED                                                                                                                                                                       [ 77%]
test_netmiko_show.py::test_clear_buffer PASSED                                                                                                                                                                              [ 83%]
test_netmiko_show.py::test_enable_mode PASSED                                                                                                                                                                               [ 88%]
test_netmiko_show.py::test_disconnect PASSED                                                                                                                                                                                [ 94%]
test_netmiko_show.py::test_disconnect_no_enable PASSED                                                                                                                                                                      [100%]

===================================================================================================== short test summary info =====================================================================================================
SKIPPED [1] /home/bspendlove/dev/git/netmiko/tests/test_netmiko_show.py:123: <Skipped instance>
SKIPPED [1] /home/bspendlove/dev/git/netmiko/tests/test_netmiko_show.py:144: TextFSM/ntc-templates not supported on this platform
SKIPPED [1] /home/bspendlove/dev/git/netmiko/tests/test_netmiko_show.py:169: Genie not supported on this platform
================================================================================================= 15 passed, 3 skipped in 52.51s ==================================================================================================
```

File: netmiko/huawei/huawei_smartax.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+from netmiko import log
+
+
+class HuaweiSmartAXSSH(CiscoBaseConnection):
+    def session_preparation(self):
+        """Prepare the session after the connection has been established."""
+        self.ansi_escape_codes = True
+        self._test_channel_read()
+        self.set_base_prompt()
+        self.disable_smart_interaction()
```
Comment (ktbyers) [inline netmiko/huawei/huawei_smartax.py]: Convert this to a private method i.e. add a leading underscore `self._disable_smart_interaction()`

File: netmiko/huawei/huawei_smartax.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+        output = string_buffer
+        pattern = rf" {code_cursor_left}"
+        output = re.sub(pattern, "", output)
+
+        log.debug("Stripping ANSI escape codes")
+        log.debug(f"new_output = {output}")
+        log.debug(f"repr = {repr(output)}")
+        return super().strip_ansi_escape_codes(output)
+
+    def disable_smart_interaction(self, command="undo smart", delay_factor=1):
```
Comment (ktbyers) [inline netmiko/huawei/huawei_smartax.py]: Convert to private method.

File: netmiko/huawei/huawei_smartax.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    def check_config_mode(self, check_string=")#"):
+        return super().check_config_mode(check_string=check_string)
+
+    def exit_config_mode(self, exit_config="return"):
+        return super().exit_config_mode(exit_config=exit_config)
+
+    def check_enable_mode(self, check_string="#"):
+        return super().check_enable_mode(check_string=check_string)
+
+    def enable(self, cmd="enable", pattern="", re_flags=re.IGNORECASE):
```
Comment (ktbyers) [inline netmiko/huawei/huawei_smartax.py]: Does this prompt for a password/secret? Usually you would have the pattern="Password" or whatever it prompts for (for the additional password)?

File: netmiko/huawei/huawei_smartax.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+        return super().enable(cmd=cmd, pattern=pattern, re_flags=re_flags)
+
+    def exit_enable_mode(self, exit_command="disable"):
+        return super().exit_enable_mode(exit_command=exit_command)
```
Comment (ktbyers) [inline netmiko/huawei/huawei_smartax.py]: Shouldn't be needed as this is identical to the CiscoBaseConnection (i.e. you can just use the method in the parent class).

File: netmiko/ssh_dispatcher.py [hunk 5]
```diff
@@ -31,7 +31,6 @@
 from netmiko.dell import DellPowerConnectSSH
 from netmiko.dell import DellPowerConnectTelnet
 from netmiko.dell import DellIsilonSSH
-from netmiko.dlink import DlinkDSTelnet, DlinkDSSSH
 from netmiko.eltex import EltexSSH, EltexEsrSSH
```
Comment (ktbyers) [inline netmiko/ssh_dispatcher.py]: It looks like you need to `git rebase` as you are removing a few things you shouldn't be removing.

File: netmiko/ssh_dispatcher.py [hunk 6]
```diff
@@ -78,8 +78,6 @@
 from netmiko.terminal_server import TerminalServerTelnet
 from netmiko.ubiquiti import UbiquitiEdgeSSH
 from netmiko.vyos import VyOSSSH
-from netmiko.watchguard import WatchguardFirewareSSH
```
Comment (ktbyers) [inline netmiko/ssh_dispatcher.py]: Need to rebase so you don't remove this.

File: netmiko/ssh_dispatcher.py [hunk 7]
```diff
@@ -117,7 +115,6 @@
     "dell_os10": DellOS10SSH,
     "dell_powerconnect": DellPowerConnectSSH,
     "dell_isilon": DellIsilonSSH,
-    "dlink_ds": DlinkDSSSH,
```
Comment (ktbyers) [inline netmiko/ssh_dispatcher.py]: shouldn't be removing this.

File: netmiko/ssh_dispatcher.py [hunk 8]
```diff
@@ -170,7 +168,6 @@
     "ubiquiti_edgeswitch": UbiquitiEdgeSSH,
     "vyatta_vyos": VyOSSSH,
     "vyos": VyOSSSH,
-    "watchguard_fireware": WatchguardFirewareSSH,
```
Comment (ktbyers) [inline netmiko/ssh_dispatcher.py]: shouldn't be removing this.

File: netmiko/ssh_dispatcher.py [hunk 9]
```diff
@@ -212,7 +209,6 @@
 CLASS_MAPPER["cisco_xr_telnet"] = CiscoXrTelnet
 CLASS_MAPPER["dell_dnos6_telnet"] = DellDNOS6Telnet
 CLASS_MAPPER["dell_powerconnect_telnet"] = DellPowerConnectTelnet
-CLASS_MAPPER["dlink_ds_telnet"] = DlinkDSTelnet
```
Comment (ktbyers) [inline netmiko/ssh_dispatcher.py]: shouldn't be removing this.

File: tests/etc/commands.yml.example [hunk 10]
```diff
@@ -282,22 +282,6 @@ keymile_nos:
...
-  extended_output: "show config current_config"   # requires paging to be disabled
-
-dlink_ds_telnet:
-  version: "show greeting_message"
-  basic: "show ipif"
-  config:
-    - enable command logging
-  config_verification: "show config current_config include \"logging\""
-  extended_output: "show config current_config"   # requires paging to be disabled
-
```
Comment (ktbyers) [inline tests/etc/commands.yml.example]: Rebase to fix this accidental removal.

File: tests/etc/responses.yml.example [hunk 11]
```diff
@@ -223,22 +223,6 @@ keymile_nos:
...
-  multiple_line_output: "End of configuration file"
-  interface_ip: 192.168.50.10
-
-dlink_ds_telnet:
-  base_prompt: DGS-3120-24TC:admin
-  router_prompt: DGS-3120-24TC:admin#
-  enable_prompt: DGS-3120-24TC:admin#
-  version_banner: "D-Link Corporation"
-  multiple_line_output: "End of configuration file"
-  interface_ip: 192.168.50.10
```
Comment (ktbyers) [inline tests/etc/responses.yml.example]: Rebase to fix this accidental removal.

File: tests/etc/test_devices.yml.example [hunk 12]
```diff
@@ -178,18 +178,6 @@ keymile_nos:
...
-  ip: 192.168.50.10
-  username: admin
-  password: admin
-
-dlink_ds_telnet:
-  device_type: dlink_ds_telnet
-  ip: 192.168.50.10
-  username: admin
-  password: admin
-
```
Comment (ktbyers) [inline tests/etc/test_devices.yml.example]: Rebase to fix.

**Comment by ktbyers** [comment]
@BSpendlove Do you know which Huawei OS(s) the standard Netmiko driver would cover? I will probably add another key in ssh_dispatcher to more clearly differentiate the current Huawei driver.

I thought you might know off the top of your head. Thanks.

**Comment by BSpendlove** [comment]
@ktbyers The current driver I'm not too sure about, I've been using it extensively on Huawei routers that run `Huawei Versatile Routing Platform Software`, I can't seem to find the actual operating system name that they create on their Multi-Access (MA) OLTs but so far I've only been able to test this current driver for the Huawei OLT MA5xxx series (OLT Series). I'll fix my mistakes and commit again haha

File: netmiko/huawei/huawei_smartax.py (hunk 3 above)
Comment (BSpendlove) [reply netmiko/huawei/huawei_smartax.py]: The enable mode for Huawei (and the same with Adtran) doesn't ask for a password, it's just a command to activate the enable mode from basic mode, should I just remove the `pattern` in that case?

File: tests/etc/test_devices.yml.example (hunk 12 above)
Comment (BSpendlove) [reply tests/etc/test_devices.yml.example]: Hey apologies, I'm still getting used to Git and forking etc.. might take me a few tries to get it back to normal 😆

File: netmiko/huawei/huawei_smartax.py (hunk 2 above)
Comment (BSpendlove) [reply netmiko/huawei/huawei_smartax.py]: Converted this as requested

File: netmiko/huawei/huawei_smartax.py (hunk 4 above)
Comment (BSpendlove) [reply netmiko/huawei/huawei_smartax.py]: I have now removed this 👍

File: netmiko/ssh_dispatcher.py (hunk 5 above)
Comment (BSpendlove) [reply netmiko/ssh_dispatcher.py]: Please check my new commit, should of fixed it

File: netmiko/ssh_dispatcher.py (hunk 8 above)
Comment (BSpendlove) [reply netmiko/ssh_dispatcher.py]: Please check my new commit, should of fixed it

File: netmiko/ssh_dispatcher.py (hunk 9 above)
Comment (BSpendlove) [reply netmiko/ssh_dispatcher.py]: Please check my new commit, should of fixed it

File: tests/etc/commands.yml.example (hunk 10 above)
Comment (BSpendlove) [reply tests/etc/commands.yml.example]: Please check my new commit, should of fixed it

File: tests/etc/responses.yml.example (hunk 11 above)
Comment (BSpendlove) [reply tests/etc/responses.yml.example]: Please check my new commit, should of fixed it

File: netmiko/huawei/huawei_smartax.py (hunk 3 above)
Comment (ktbyers) [reply netmiko/huawei/huawei_smartax.py]: The way you did it is fine (given the way Huawei operates)...just wanted to verify it.

**Comment by ktbyers** [comment]
@BSpendlove Good work on this...thanks for the contribution :-)

**Pull request merged**